File: Dynamic_Programming/1137.N-th_Tribonacci_Number/Python/1137.N-th_Tribonacci_Number.py
class Solution(object):
    def tribonacci(self, n):
        """
        :type n: int
        :rtype: int
        """
        #======================================================#
        # Dynamic-programming (i.e. DP) based traversal method #
        #======================================================#
        if (not n): #Issue/Boundary-case handling
            return 0
        
        if (n <= 2): #Issue/Boundary-case handling
            return 1


        ############
        #Initialize
        ##### Record dynamic-programming array #####
        record_dp = [0 for _ in range((n + 1))]

        ####################
        #Whole process/flow
        record_dp[0] = 0 #Update/Overwrite
        record_dp[1] = 1 #Update/Overwrite
        record_dp[2] = 1 #Update/Overwrite

        for n_idx in range(3, (n + 1)):
            record_dp[n_idx] = (record_dp[(n_idx - 1)] + record_dp[(n_idx - 2)] + record_dp[n_idx - 3]) #Keep updating/overwriting

        return record_dp[n]

File: Dynamic_Programming/1137.N-th_Tribonacci_Number/C++/1137.N-th_Tribonacci_Number.cpp
class Solution {
public:
    int tribonacci(int n) 
    {
        //================================//
        // Input type:                    //
        //   - int (n-th times)           //
        // Return type:                   //
        //   - int (result summary value) //
        //================================//
        if (! n) { return 0; } //Issue, Boundary-case handling
        
        if (n <= 2) { return 1; } //Issue, Boundary-case handling


        /*Initialize*/
        ///// Record dynamic-programming (i.e. DP) vector /////
        vector<int> record_dp((n + 1), 0);

        /*Dynamic-programming based loop traversal*/
        record_dp[0] = 0; record_dp[1] = 1; record_dp[2] = 1; //Update, Overwrite

        for (int n_idx = 3; (n_idx <= n); (n_idx++)) //N-th times
        {
            record_dp[n_idx] = (record_dp[(n_idx - 1)] + record_dp[(n_idx - 2)] + record_dp[(n_idx - 3)]); //Keep updating, overwriting

        } //N-th times

        return record_dp[n];
    }
};
